src/compiler/checker.ts

```diff
@@ -9082,15 +9082,15 @@ namespace ts {
                         return getInferredClassType(classSymbol);
                     }
                 }
-                const type = getContextuallyTypedThisType(container);
-                if (type) {
-                    return type;
-                }
 
                 const thisType = getThisTypeOfDeclaration(container);
                 if (thisType) {
                     return thisType;
                 }
+                const type = getContextuallyTypedThisType(container);

```

Comment:
Based on our in-person discussion, I checked on contextual typing for parameters. It's 1 of 9 cases in `getTypeForVariableLikeDeclaration`: 

```ts
                const type = declaration.symbol.name === "this"
                    ? getContextuallyTypedThisType(func)
                    : getContextuallyTypedParameterType(<ParameterDeclaration>declaration);
```

The contextual typing for an undeclared `this` is basically the same thing here in `checkThisExpression` since there is no parameter to get the type of.